state-and-event-handling/event-handling.md

```diff
@@ -0,0 +1,182 @@
+
+Every time we click our "Like" button, we'll see our message in the console!
+
+```
+Hello! We're in printMessage!
+```
+
+### Event Handling Functions Can Accept `event`
+
+We can get details about our events in our event-handling functions.
+
+If our event-handling function accepts arguments, the first argument will be an event object.
+
+Consider this code, where `printMessage` accepts an event object, and then prints it.
+
+<!-- prettier-ignore-start -->
+```js
+const Post = () => {
+    const printMessage = (event) => {
+        console.log('Hello! We\'re in printMessage!');
+        console.log('This event object contains details about the event:', event);
+    }
+
+    return (<section>
+        <button onClick={printMessage}>Like</button>
+    </section>)
+}
+```
+<!-- prettier-ignore-end -->
+
+This is what prints to the browser console. (Your results may vary depending on your browser.)
+
+```
+Hello! We're in printMessage!
+This event object contains details about the event: 
+Object { _reactName: "onClick", ...
+```
+
+### Event Handling Functions Are Functions
+
+It's helpful to know that our event handling functions are, otherwise, normal functions.
+
+They can call other functions inside them, and pass arguments to them, too.
+
+Consider this `Post` component. It:
+
+- Defines a function, `printName`, which takes in a `name` argument
+- Defines another function, `printMessage`. It calls `printName` with the argument `"Ada Lovelace"`.
+- Sets `printMessage` as the click event handler for a "Like" button
+
+<!-- prettier-ignore-start -->
+```js
+const Post = () => {
+    const printName = (name) => {
+        console.log('We\'re in printName!');
+        console.log(`Hello, ${name}!`);
+    }
+
+    const printMessage = () => {
+        printName('Ada Lovelace');
+        console.log('Now, we\'re in printMessage!');
+    }
+
+    return (<section>
+        <button onClick={printMessage}>Like</button>
+    </section>)
+}
+```
+<!-- prettier-ignore-end -->
+
+Our output proves that our `printMessage` event handler can invoke other functions.
+
+```
+We're in printName!
+Hello, Ada Lovelace!
+Now, we're in printMessage!
+```
+
+## Pattern: Anonymous Functions
+
+Our event handlers can be _anonymous functions_.
+
+Consider our `printMessage` and `printName` example. Let's refactor our `printMessage` into an inline, anonymous function:
+
+<!-- prettier-ignore-start -->
+```js
+const Post = () => {
+    const printName = (name) => {
+        console.log('We\'re in printName!');
+        console.log(`Hello, ${name}!`);
+    }
+
+    return (<section>
+        <button onClick={() => printName('Ada Lovelace')}>Like</button>
+    </section>)
+}
+```
+<!-- prettier-ignore-end -->
+
+This code produces the output:
+
+```
+We're in printName!
+Hello, Ada Lovelace!
+```
+
+Every time we click on our "like" button, our anonymous function will fire. This, in turn, invokes `printName('Ada Lovelace')`!
+
+This is a common pattern in React code.
```

Comment:
I don't think this is clear about the pattern. Let's be a little more explicit and warn about a common error students run into.
```suggestion
Using anonymous functions for event handlers is a common pattern in React, especially when the function we want to call needs us to pass in some parameters.

A common mistake when trying to call a function that takes an argument as an event handler is to try to write the function call, in this case `printName('Ada Lovelace')`, directly as the even handler.

But this does not have the desired result! Instead, it:
1. Calls `printName('Ada Lovelace')` as the button is being rendered rather than when it is clicked. This causes the output message to appear while the page is still being rendered. Not what we wanted!
1. Sets the _result_ of calling `printName('Ada Lovelace')` as the click handler. Since `printName` returns nothing, the result is `undefined`. So no event handler is registered for `onClick`, and the button does nothing when it is clicked. Also not what we wanted!
```